SolvePnPRansac Fails because calls to solvePnP with less than 6 points

##### System information (version)
<!-- Example
- OpenCV => 4.3
- Operating System / Platform => Linux 64 Bit
- Compiler => gcc
-->

- OpenCV => :grey_question:
- Operating System / Platform => :grey_question:
- Compiler => :grey_question:

##### Detailed description

<!-- your description -->
solvePnPRansac calls solvePnP wihch calls cvFIndExtrinsicCameraParams2 using less than 6 points.
Then, the following error is raised:
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.3.0) /home/salinas/Libraries/opencv/opencv-4.3.0/modules/calib3d/src/calibration.cpp:1173: error: (-2:Unspecified error) in function 'void cvFindExtrinsicCameraParams2(const CvMat*, const CvMat*, const CvMat*, const CvMat*, CvMat*, CvMat*, int)'
> DLT algorithm needs at least 6 points for pose estimation from 3D-2D point correspondences. (expected: 'count >= 6'), where
>     'count' is 5
> must be greater than or equal to
>     '6' is 6

##### Steps to reproduce

<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   <!--
   OpenCV team works with answers.opencv.org, Stack Overflow and other communities
   to discuss problems. Tickets with question without real issue statement will be
   closed.
   -->
 - [ ] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   <!--
   Places to check:
   * OpenCV documentation: https://docs.opencv.org
   * FAQ page: https://github.com/opencv/opencv/wiki/FAQ
   * OpenCV forum: https://answers.opencv.org
   * OpenCV issue tracker: https://github.com/opencv/opencv/issues?q=is%3Aissue
   * Stack Overflow branch: https://stackoverflow.com/questions/tagged/opencv
   -->
 - [x] I updated to latest OpenCV version and the issue is still there
   <!--
   master branch for OpenCV 4.x and 3.4 branch for OpenCV 3.x releases.
   OpenCV team supports only latest release for each branch.
   The ticket is closed, if the problem is not reproduced with modern version.
   -->
 - [ ] There is reproducer code and related data files: videos, images, onnx, etc
   <!--
   The best reproducer -- test case for OpenCV that we can add to the library.
   Recommendations for media files and binary files:
   * Try to reproduce the issue with images and videos in opencv_extra repository
     to reduce attachment size
   * Use PNG for images, if you report some CV related bug, but not image reader
     issue
   * Attach the image as archite to the ticket, if you report some reader issue.
     Image hosting services compress images and it breaks the repro code.
   * Provide ONNX file for some public model or ONNX file with with random weights,
     if you report ONNX parsing or handling issue. Architecture details diagram
     from netron tool can be very useful too. See https://lutzroeder.github.io/netron/
   -->

--- a/modules/calib3d/src/solvepnp.cpp
+++ b/modules/calib3d/src/solvepnp.cpp
@@ -311,19 +311,26 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
 
     opoints_inliers.resize(npoints1);
     ipoints_inliers.resize(npoints1);
-    result = solvePnP(opoints_inliers, ipoints_inliers, cameraMatrix,
-                      distCoeffs, rvec, tvec, useExtrinsicGuess,
-                      (flags == SOLVEPNP_P3P || flags == SOLVEPNP_AP3P) ? SOLVEPNP_EPNP : flags) ? 1 : -1;
+    try
+    {
+        result = solvePnP(opoints_inliers, ipoints_inliers, cameraMatrix,
+                          distCoeffs, rvec, tvec, useExtrinsicGuess,
+                          (flags == SOLVEPNP_P3P || flags == SOLVEPNP_AP3P) ? SOLVEPNP_EPNP : flags) ? 1 : -1;
+    }
+    catch (const cv::Exception& e)
+    {
+        result = 0;
+        CV_UNUSED(e);
+        CV_LOG_DEBUG(NULL, "solvePnPRansac(): error occurred in solvePnP stage:\n" << e.what());
+    }
 
-    if( result <= 0 )
+    if(result <= 0)
     {
         _rvec.assign(_local_model.col(0));    // output rotation vector
         _tvec.assign(_local_model.col(1));    // output translation vector
 
-        if( _inliers.needed() )
-            _inliers.release();
-
-        return false;
+        CV_LOG_DEBUG(NULL, "solvePnPRansac(): solvePnP stage to compute the final pose using points "
+            "in the consensus set failed, use result from MSS (Minimal Sample Sets) stage instead.");
     }
     else
     {

--- a/modules/calib3d/src/solvepnp.cpp
+++ b/modules/calib3d/src/solvepnp.cpp
@@ -319,18 +319,35 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
     }
     catch (const cv::Exception& e)
     {
-        result = 0;
-        CV_UNUSED(e);
-        CV_LOG_DEBUG(NULL, "solvePnPRansac(): error occurred in solvePnP stage:\n" << e.what());
+        if (flags == SOLVEPNP_ITERATIVE &&
+            npoints1 == 5 &&
+            e.what() &&
+            std::string(e.what()).find("DLT algorithm needs at least 6 points") != std::string::npos
+        )
+        {
+            CV_LOG_INFO(NULL, "solvePnPRansac(): solvePnP stage to compute the final pose using points "
+                "in the consensus set raised DLT 6 points exception, use result from MSS (Minimal Sample Sets) stage instead.");
+            rvec = _local_model.col(0);    // output rotation vector
+            tvec = _local_model.col(1);    // output translation vector
+            result = 1;
+        }
+        else
+        {
+            // raise other exceptions
+            throw;
+        }
     }
 
-    if(result <= 0)
+    if (result <= 0)
     {
         _rvec.assign(_local_model.col(0));    // output rotation vector
         _tvec.assign(_local_model.col(1));    // output translation vector
 
-        CV_LOG_DEBUG(NULL, "solvePnPRansac(): solvePnP stage to compute the final pose using points "
-            "in the consensus set failed, use result from MSS (Minimal Sample Sets) stage instead.");
+        if (_inliers.needed())
+            _inliers.release();
+
+        CV_LOG_DEBUG(NULL, "solvePnPRansac(): solvePnP stage to compute the final pose using points in the consensus set failed. Return false");
+        return false;
     }
     else
     {

--- a/modules/calib3d/test/test_solvepnp_ransac.cpp
+++ b/modules/calib3d/test/test_solvepnp_ransac.cpp
@@ -837,6 +837,39 @@ TEST(Calib3d_SolvePnPRansac, double_support)
     EXPECT_LE(cvtest::norm(t, Mat_<double>(tF), NORM_INF), 1e-3);
 }
 
+TEST(Calib3d_SolvePnPRansac, bad_input_points)
+{
+    // with this specific data
+    // when computing the final pose using points in the consensus set with SOLVEPNP_ITERATIVE and solvePnP()
+    // an exception is thrown from solvePnP because there are 5 non-coplanar 3D points and the DLT algorithm needs at least 6 non-coplanar 3D points
+    // with PR #19253 we choose to return true, with the pose estimated from the MSS stage instead of throwing the exception
+
+    Mat pts2d = (Mat_<float>(6, 2) <<
+        -5.38358629e-01, -5.09638414e-02,
+        -5.07192254e-01, -2.20743284e-01,
+        -5.43107152e-01, -4.90474701e-02,
+        -5.54325163e-01, -1.86715424e-01,
+        -5.59334219e-01, -4.01909500e-02,
+        -5.43504596e-01, -4.61776406e-02);
+
+    Mat pts3d = (Mat_<float>(6, 3) <<
+        -3.01153604e-02, -1.55665115e-01, 4.50000018e-01,
+        4.27827090e-01, 4.28645730e-01, 1.08600008e+00,
+        -3.14165242e-02, -1.52656138e-01, 4.50000018e-01,
+        -1.46217480e-01, 5.57961613e-02, 7.17000008e-01,
+        -4.89348806e-02, -1.38795510e-01, 4.47000027e-01,
+        -3.13065052e-02, -1.52636901e-01, 4.51000035e-01);
+
+    Mat camera_mat = Mat::eye(3, 3, CV_64FC1);
+    Mat rvec, tvec;
+    vector<int> inliers;
+
+    // solvePnPRansac will return true with 5 inliers, which means the result is from MSS stage.
+    bool result = solvePnPRansac(pts3d, pts2d, camera_mat, noArray(), rvec, tvec, false, 100, 4.f / 460.f, 0.99, inliers);
+    EXPECT_EQ(inliers.size(), size_t(5));
+    EXPECT_TRUE(result);
+}
+
 TEST(Calib3d_SolvePnP, input_type)
 {
     Matx33d intrinsics(5.4794130238156129e+002, 0., 2.9835545700043139e+002, 0.,

--- a/modules/calib3d/test/test_solvepnp_ransac.cpp
+++ b/modules/calib3d/test/test_solvepnp_ransac.cpp
@@ -837,28 +837,32 @@ TEST(Calib3d_SolvePnPRansac, double_support)
     EXPECT_LE(cvtest::norm(t, Mat_<double>(tF), NORM_INF), 1e-3);
 }
 
-TEST(Calib3d_SolvePnPRansac, bad_input_points)
+TEST(Calib3d_SolvePnPRansac, bad_input_points_19253)
 {
     // with this specific data
     // when computing the final pose using points in the consensus set with SOLVEPNP_ITERATIVE and solvePnP()
     // an exception is thrown from solvePnP because there are 5 non-coplanar 3D points and the DLT algorithm needs at least 6 non-coplanar 3D points
     // with PR #19253 we choose to return true, with the pose estimated from the MSS stage instead of throwing the exception
 
-    Mat pts2d = (Mat_<float>(6, 2) <<
-        -5.38358629e-01, -5.09638414e-02,
-        -5.07192254e-01, -2.20743284e-01,
-        -5.43107152e-01, -4.90474701e-02,
-        -5.54325163e-01, -1.86715424e-01,
-        -5.59334219e-01, -4.01909500e-02,
-        -5.43504596e-01, -4.61776406e-02);
+    float pts2d_[] = {
+        -5.38358629e-01f, -5.09638414e-02f,
+        -5.07192254e-01f, -2.20743284e-01f,
+        -5.43107152e-01f, -4.90474701e-02f,
+        -5.54325163e-01f, -1.86715424e-01f,
+        -5.59334219e-01f, -4.01909500e-02f,
+        -5.43504596e-01f, -4.61776406e-02f
+    };
+    Mat pts2d(6, 2, CV_32FC1, pts2d_);
 
-    Mat pts3d = (Mat_<float>(6, 3) <<
-        -3.01153604e-02, -1.55665115e-01, 4.50000018e-01,
-        4.27827090e-01, 4.28645730e-01, 1.08600008e+00,
-        -3.14165242e-02, -1.52656138e-01, 4.50000018e-01,
-        -1.46217480e-01, 5.57961613e-02, 7.17000008e-01,
-        -4.89348806e-02, -1.38795510e-01, 4.47000027e-01,
-        -3.13065052e-02, -1.52636901e-01, 4.51000035e-01);
+    float pts3d_[] = {
+        -3.01153604e-02f, -1.55665115e-01f, 4.50000018e-01f,
+        4.27827090e-01f, 4.28645730e-01f, 1.08600008e+00f,
+        -3.14165242e-02f, -1.52656138e-01f, 4.50000018e-01f,
+        -1.46217480e-01f, 5.57961613e-02f, 7.17000008e-01f,
+        -4.89348806e-02f, -1.38795510e-01f, 4.47000027e-01f,
+        -3.13065052e-02f, -1.52636901e-01f, 4.51000035e-01f
+    };
+    Mat pts3d(6, 3, CV_32FC1, pts3d_);
 
     Mat camera_mat = Mat::eye(3, 3, CV_64FC1);
     Mat rvec, tvec;

